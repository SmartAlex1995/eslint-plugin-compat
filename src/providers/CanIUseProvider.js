// @flow
// $FlowFixMe: Flow import error
import caniuseRecord from 'caniuse-db/fulldata-json/data-2.0.json'; // eslint-disable-line
import type { Node, ESLintNode, Targets, Target } from '../LintTypes';


type TargetMetadata = {
  // The list of targets supported by the provider
  targets: Targets
};

type CanIUseRecord = {
  data: {
    [browser: string]: {
      [version: string]: string
    }
  }
};

// HACK: modern targets should be determined once at runtime
export const targetMetadata: TargetMetadata = {
  targets: [
    'chrome', 'firefox', 'opera', 'safari', 'ie', 'edge', 'ios_saf',
    'op_mini', 'android', 'bb', 'op_mob', 'and_chr', 'and_ff', 'ie_mob', 'and_uc',
    'samsung', 'baidu'
  ]
};

const targetNameMappings = {
  chrome: 'Chrome',
  firefox: 'Firefox',
  opera: 'Opera',
  baidu: 'Baidu',
  and_qq: 'QQ Browser',
  safari: 'Safari',
  android: 'Android Browser',
  ie: 'IE',
  edge: 'Edge',
  ios_saf: 'iOS Safari',
  op_mini: 'Opera Mini',
  bb: 'Blackberry Browser',
  op_mob: 'Opera Mobile',
  and_chr: 'Android Chrome',
  and_ff: 'Android Firefox',
  ie_mob: 'IE Mobile',
  and_uc: 'Android UC Browser',
  samsung: 'Samsung Browser'
};

/**
 * Take a target's id and return it's full name by using `targetNameMappings`
 * ex. {target: and_ff, version: 40} => 'Android FireFox 40'
 */
function formatTargetNames(target: Target): string {
  return `${targetNameMappings[target.target]} ${target.version}`;
}

/**
 * Check version for the range format.
 * ex. 10.0-10.2
 */
function versionIsRange(version: string): bool {
  return version.includes('-');
}

/**
 * Parse version from caniuse and compare with parsed version from browserslist.
 */
function compareRanges(targetVersion: number, statsVersion: string): bool {
  return targetVersion === parseFloat(statsVersion);
}

/**
 * Return an array of all unsupported targets
 */
export function getUnsupportedTargets(node: Node, targets: Targets): Array<string> {
  // Check the CanIUse database to see if targets are supported
  const { stats } = (caniuseRecord: CanIUseRecord).data[node.id];

  return targets.filter((target: Target): bool => {
    const { version } = target;
    const targetStats = stats[target.target];

    return (versionIsRange(version))
      ? Object.keys(targetStats).some((statsVersion: string): bool =>
        ((versionIsRange(statsVersion) && compareRanges(target.parsedVersion, statsVersion))
          ? !targetStats[statsVersion].includes('y')
          : false))
      : targetStats[version] && !targetStats[version].includes('y');
  })
    .map(formatTargetNames);
}

function isValid(node: Node, eslintNode: ESLintNode, targets: Targets): bool {
  // Filter non-matching objects and properties
  switch (eslintNode.type) {
    case 'CallExpression':
      if (!eslintNode.callee) return true;
      if (
        eslintNode.callee.name !== node.object
      ) return true;
      break;
    case 'NewExpression':
      if (!eslintNode.callee) return true;
      if (
        eslintNode.callee.name !== node.object
      ) return true;
      break;
    case 'MemberExpression':
      // Pass tests if non-matching object or property
      if (!eslintNode.object || !eslintNode.property) return true;
      if (eslintNode.object.name !== node.object) return true;

      // If the property is missing from the rule, it means that only the
      // object is required to determine compatibility
      if (!node.property) break;

      if (eslintNode.property.name !== node.property) return true;
      break;
    default:
      return true;
  }

  return getUnsupportedTargets(node, targets).length === 0;
}

//
// TODO: Migrate to compat-db
// TODO: Refactor isValid(), remove from rules
//

const CanIUseProvider: Array<Node> = [
  // new ServiceWorker()
  {
    id: 'serviceworkers',
    ASTNodeType: 'NewExpression',
    object: 'ServiceWorker',
    isValid,
    getUnsupportedTargets
  },
  {
    id: 'serviceworkers',
    ASTNodeType: 'MemberExpression',
    object: 'navigator',
    property: 'serviceWorker',
    isValid,
    getUnsupportedTargets
  },
  // document.querySelector()
  {
    id: 'queryselector',
    ASTNodeType: 'MemberExpression',
    object: 'document',
    property: 'querySelector',
    isValid,
    getUnsupportedTargets
  },
  // WebAssembly
  {
    id: 'wasm',
    ASTNodeType: 'MemberExpression',
    object: 'WebAssembly',
    isValid,
    getUnsupportedTargets
  },
  // IntersectionObserver
  {
    id: 'intersectionobserver',
    ASTNodeType: 'NewExpression',
    object: 'IntersectionObserver',
    isValid,
    getUnsupportedTargets
  },
  // PaymentRequest
  {
    id: 'payment-request',
    ASTNodeType: 'NewExpression',
    object: 'PaymentRequest',
    isValid,
    getUnsupportedTargets
  },
  // Promises
  {
    id: 'promises',
    ASTNodeType: 'NewExpression',
    object: 'Promise',
    isValid,
    getUnsupportedTargets
  },
  {
    id: 'promises',
    ASTNodeType: 'MemberExpression',
    object: 'Promise',
    property: 'resolve',
    isValid,
    getUnsupportedTargets
  },
  {
    id: 'promises',
    ASTNodeType: 'MemberExpression',
    object: 'Promise',
    property: 'all',
    isValid,
    getUnsupportedTargets
  },
  {
    id: 'promises',
    ASTNodeType: 'MemberExpression',
    object: 'Promise',
    property: 'race',
    isValid,
    getUnsupportedTargets
  },
  {
    id: 'promises',
    ASTNodeType: 'MemberExpression',
    object: 'Promise',
    property: 'reject',
    isValid,
    getUnsupportedTargets
  },
  // fetch
  {
    id: 'fetch',
    ASTNodeType: 'CallExpression',
    object: 'fetch',
    isValid,
    getUnsupportedTargets
  },
  // document.currentScript()
  {
    id: 'document-currentscript',
    ASTNodeType: 'MemberExpression',
    object: 'document',
    property: 'currentScript',
    isValid,
    getUnsupportedTargets
  },
  // URL
  {
    id: 'url',
    ASTNodeType: 'NewExpression',
    object: 'URL',
    isValid,
    getUnsupportedTargets
  },
  // URLSearchParams
  {
    id: 'urlsearchparams',
    ASTNodeType: 'NewExpression',
    object: 'URLSearchParams',
    isValid,
    getUnsupportedTargets
  }
];

export default CanIUseProvider;
