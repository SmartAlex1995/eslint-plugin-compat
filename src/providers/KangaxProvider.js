// @flow
import type { Node } from '../LintTypes';


const KangaxProvider: Array<Node> = [];

export default KangaxProvider;
