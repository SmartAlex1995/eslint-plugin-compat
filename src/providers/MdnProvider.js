import AstMetadata from 'ast-metadata-inferer';
import semver from 'semver';
import type { Node, Targets, Target } from '../LintTypes';

type AstMetadataRecordType = {
  apiType: 'js-api' | 'css-api',
  type: 'js-api' | 'css-api',
  protoChain: Array<string>,
  protoChainId: string,
  astNodeTypes: Array<string>,
  isStatic: boolean,
  compat: {
    support: {
      [browserName: string]: {
        // If a version is true then it is supported but version is unsure
        version_added: string | boolean
      }
    },
    [x: string]: any
  }
};

const mdnRecords: Map<string, AstMetadataRecordType> = new Map(
  AstMetadata.map(e => [e.protoChainId, e])
);

/**
 * Map ids of mdn targets to their "common/friendly" name
 */
const targetNameMappings = {
  chrome: 'Chrome',
  firefox: 'Firefox',
  opera: 'Opera',
  safari: 'Safari',
  ie: 'IE',
  edge: 'Edge',
  safari_ios: 'iOS Safari',
  opera_android: 'Opera Mobile',
  chrome_android: 'Android Chrome',
  edge_mobile: 'Edge Mobile',
  firefox_android: 'Android Firefox',
  webview_android: 'WebView Android',
  samsunginternet_android: 'Samsung Browser',
  nodes: 'Node.js'
};

/**
 * Take a target's id and return it's full name by using `targetNameMappings`
 * ex. {target: and_ff, version: 40} => 'Android FireFox 40'
 */
function formatTargetNames(target: Target): string {
  return `${targetNameMappings[target.target]} ${target.version}`;
}

/**
 * Convert '9' => '9.0.0'
 */
function customCoerce(version: string): string {
  return version.length === 1 ? [version, 0, 0].join('.') : version;
}

/*
 * Return if MDN supports the API or not
 */
export function mdnSupported(node: Node, { version, target }: Target): boolean {
  // If no record could be found, return false. Rules might not
  // be found because they could belong to another provider
  if (!mdnRecords.has(node.protoChainId)) return true;
  const record = mdnRecords.get(node.protoChainId);
  if (!record || !record.compat.support) return true;
  const compatRecord = record.compat.support[target];
  if (!compatRecord) return true;
  if (!Array.isArray(compatRecord) && !('version_added' in compatRecord))
    return true;
  const { version_added: versionAdded } = Array.isArray(compatRecord)
    ? compatRecord.find(e => 'version_added' in e)
    : compatRecord;

  // If a version is true then it is supported but version is unsure
  if (typeof versionAdded === 'boolean') return versionAdded;
  if (versionAdded === null) return true;

  // Special case for Safari TP: TP is always gte than any other releases
  if (target === 'safari') {
    if (version === 'TP') return true;
    if (versionAdded === 'TP') return false;
  }
  // A browser supports an API if its version is greater than or equal
  // to the first version of the browser that API was added in
  const semverCurrent = semver.coerce(customCoerce(version));
  const semverAdded = semver.coerce(customCoerce(versionAdded));

  // semver.coerce() might be null for non-semvers (other than Safari TP)
  // Just warn and treat features as supported here for now to avoid lint from
  // crashing
  if (!semverCurrent) {
    // eslint-disable-next-line no-console
    console.warn(
      `eslint-plugin-compat: A non-semver target "${target} ${version}" matched for the feature ${node.protoChainId}, skipping. You're welcome to submit this log to https://github.com/amilajack/eslint-plugin-compat/issues for analysis.`
    );
    return true;
  }
  if (!versionAdded) {
    // eslint-disable-next-line no-console
    console.warn(
      `eslint-plugin-compat: The feature ${node.protoChainId} is supported since a non-semver target "${target} ${versionAdded}", skipping. You're welcome to submit this log to https://github.com/amilajack/eslint-plugin-compat/issues for analysis.`
    );
    return true;
  }
  return semver.gte(semverCurrent, semverAdded);
}

/**
 * Return an array of all unsupported targets
 */
export function getUnsupportedTargets(
  node: Node,
  targets: Targets
): Array<string> {
  return targets
    .filter(target => !mdnSupported(node, target))
    .map(formatTargetNames);
}

function getMetadataName(metadata: Node) {
  switch (metadata.protoChain.length) {
    case 1: {
      return metadata.protoChain[0];
    }
    default:
      return `${metadata.protoChain.join('.')}()`;
  }
}

const MdnProvider: Array<Node> = AstMetadata
  // Create entries for each ast node type
  .map(metadata =>
    metadata.astNodeTypes.map(astNodeType => ({
      ...metadata,
      name: getMetadataName(metadata),
      id: metadata.protoChainId,
      protoChainId: metadata.protoChainId,
      astNodeType,
      object: metadata.protoChain[0],
      // @TODO Handle cases where 'prototype' is in protoChain
      property: metadata.protoChain[1]
    }))
  )
  // Flatten the array of arrays
  .reduce((p, c) => [...p, ...c])
  // Add rule and target support logic for each entry
  .map(rule => ({
    ...rule,
    getUnsupportedTargets
  }));

export default MdnProvider;
