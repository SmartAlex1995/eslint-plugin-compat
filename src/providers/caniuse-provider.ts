import canIUseRecords from "caniuse-db/fulldata-json/data-2.0.json";
import { STANDARD_TARGET_NAME_MAPPING } from "../constants";
import { AstMetadataApiWithUnsupportedTargets, Target } from "../types";

type CanIUseStats = {
  [browser: string]: {
    [version: string]: string;
  };
};

type CanIUseRecords = {
  data: CanIUseStats;
};

/**
 * Take a target's id and return it's full name by using `targetNameMappings`
 * ex. {target: and_ff, version: 40} => 'Android FireFox 40'
 */
function formatTargetNames(target: Target): string {
  const name = STANDARD_TARGET_NAME_MAPPING[target.target] || target.target;
  return `${name} ${target.version}`;
}

/**
 * Check if a browser version is in the range format
 * ex. 10.0-10.2
 */
function versionIsRange(version: string): boolean {
  return version.includes("-");
}

/**
 * Parse version from caniuse and compare with parsed version from browserslist.
 */
function areVersionsEqual(
  targetVersion: number,
  statsVersion: string
): boolean {
  return targetVersion === parseFloat(statsVersion);
}

/*
 * Check the CanIUse database to see if targets are supported
 *
 * If no record could be found, return true. Rules might not
 * be found because they could belong to another provider
 */
function isSupportedByCanIUse(
  node: AstMetadataApiWithUnsupportedTargets,
  { version, target, parsedVersion }: Target
): boolean {
  const data = (canIUseRecords as CanIUseRecords).data[node.caniuseId];

  if (!data) return true;
  const { stats } = data;
  if (!(target in stats)) return true;

  const targetStats = stats[target];

  if (typeof version === "string" && versionIsRange(version)) {
    return Object.keys(targetStats).some((statsVersion: string): boolean =>
      versionIsRange(statsVersion) &&
      areVersionsEqual(parsedVersion, statsVersion)
        ? !targetStats[statsVersion].includes("y")
        : true
    );
  }

  // @TODO: This assumes that all versions are included in the cainuse db. If this is incorrect,
  //        this will return false negatives. To properly do this, we have to to range comparisons.
  //        Ex. given query for 50 and only version 40 exists in db records, return true
  if (!(version in targetStats)) return true;
  if (!targetStats[version]) return true;

  return targetStats[version].includes("y");
}

/**
 * Return an array of all unsupported targets
 */
export function getUnsupportedTargets(
  node: AstMetadataApiWithUnsupportedTargets,
  targets: Target[]
): string[] {
  return targets
    .filter((target) => !isSupportedByCanIUse(node, target))
    .map(formatTargetNames);
}

const CanIUseProvider: Array<AstMetadataApiWithUnsupportedTargets> = [
  // new ServiceWorker()
  {
    caniuseId: "serviceworkers",
    astNodeType: "NewExpression",
    object: "ServiceWorker",
  },
  {
    caniuseId: "serviceworkers",
    astNodeType: "MemberExpression",
    object: "navigator",
    property: "serviceWorker",
  },
  // document.querySelector()
  {
    caniuseId: "queryselector",
    astNodeType: "MemberExpression",
    object: "document",
    property: "querySelector",
  },
  // IntersectionObserver
  {
    caniuseId: "intersectionobserver",
    astNodeType: "NewExpression",
    object: "IntersectionObserver",
  },
  // ResizeObserver
  {
    caniuseId: "resizeobserver",
    astNodeType: "NewExpression",
    object: "ResizeObserver",
  },
  // PaymentRequest
  {
    caniuseId: "payment-request",
    astNodeType: "NewExpression",
    object: "PaymentRequest",
  },
  // Promises
  {
    caniuseId: "promises",
    astNodeType: "NewExpression",
    object: "Promise",
  },
  {
    caniuseId: "promises",
    astNodeType: "MemberExpression",
    object: "Promise",
    property: "resolve",
  },
  {
    caniuseId: "promises",
    astNodeType: "MemberExpression",
    object: "Promise",
    property: "all",
  },
  {
    caniuseId: "promises",
    astNodeType: "MemberExpression",
    object: "Promise",
    property: "race",
  },
  {
    caniuseId: "promises",
    astNodeType: "MemberExpression",
    object: "Promise",
    property: "reject",
  },
  // fetch
  {
    caniuseId: "fetch",
    astNodeType: "CallExpression",
    object: "fetch",
  },
  // document.currentScript()
  {
    caniuseId: "document-currentscript",
    astNodeType: "MemberExpression",
    object: "document",
    property: "currentScript",
  },
  // URL
  {
    caniuseId: "url",
    astNodeType: "NewExpression",
    object: "URL",
  },
  // URLSearchParams
  {
    caniuseId: "urlsearchparams",
    astNodeType: "NewExpression",
    object: "URLSearchParams",
  },
  // performance.now()
  {
    caniuseId: "high-resolution-time",
    astNodeType: "MemberExpression",
    object: "performance",
    property: "now",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "TypedArray",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Int8Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Uint8Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Uint8ClampedArray",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Int16Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Uint16Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Int32Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Uint32Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Float32Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: "NewExpression",
    object: "Float64Array",
  },
].map((rule) => ({
  ...rule,
  getUnsupportedTargets,
  id: rule.property ? `${rule.object}.${rule.property}` : rule.object,
  protoChainId: rule.property ? `${rule.object}.${rule.property}` : rule.object,
  protoChain: rule.property ? [rule.object, rule.property] : [rule.object],
}));

export default CanIUseProvider;
