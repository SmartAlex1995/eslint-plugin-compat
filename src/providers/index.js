// @flow
import canIUseNodes from "./CanIUseProvider";
import mdnNodes from "./MdnProvider";
import type { Node } from "../LintTypes";

// eslint-disable-next-line import/prefer-default-export
export const nodes: Array<Node> = [...canIUseNodes, ...mdnNodes];
