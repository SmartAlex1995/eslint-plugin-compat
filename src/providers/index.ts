/*
 * Step 3) Compat use CanIUse and MDN providers to check if a target browser supports a particular API
 */
import canIUseNodes from "./caniuse-provider";
import mdnNodes from "./mdn-provider";
import type { AstMetadataApiWithTargetsResolver } from "../types";

// eslint-disable-next-line import/prefer-default-export
export const nodes: Array<AstMetadataApiWithTargetsResolver> = [
  ...canIUseNodes,
  ...mdnNodes,
];
