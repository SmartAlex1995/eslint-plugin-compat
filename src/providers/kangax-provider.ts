import { Node } from "../types";

// TODO: not implemented yet.
const KangaxProvider: Array<Node> = [];

export default KangaxProvider;
