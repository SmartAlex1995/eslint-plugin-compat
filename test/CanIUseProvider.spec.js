import DetermineTargetsFromConfig, { Versioning } from '../src/Versioning';
import { getUnsupportedTargets } from '../src/providers/CanIUseProvider';
import expectRangeResultJSON from './expect-range-result-config.json';


describe('CanIUseProvider', () => {
  it('should return unsupported ios targets with range value for Fetch API', () => {
    const node = { id: 'fetch' };
    const config = DetermineTargetsFromConfig(expectRangeResultJSON.browsers);
    const targets = Versioning(config);
    const result = getUnsupportedTargets(node, targets);
    expect(result).toMatchSnapshot();
  });
});
