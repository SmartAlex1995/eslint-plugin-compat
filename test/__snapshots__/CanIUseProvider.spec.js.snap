// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanIUseProvider should return unsupported ios targets with range value for Fetch API 1`] = `
Array [
  "iOS Safari 10.0-10.2",
]
`;
